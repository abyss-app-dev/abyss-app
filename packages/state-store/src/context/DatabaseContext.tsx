import type { SQliteClient } from '@abyss/records';
import { createContext, type ReactNode, useContext } from 'react';

interface DatabaseContextType {
    database: SQliteClient | null;
}

const DatabaseContext = createContext<DatabaseContextType>({ database: null });

interface DatabaseProviderProps {
    database: SQliteClient;
    children: ReactNode;
}

export function DatabaseProvider({ database, children }: DatabaseProviderProps) {
    return <DatabaseContext.Provider value={{ database }}>{children}</DatabaseContext.Provider>;
}

export function useDatabase(): SQliteClient {
    const context = useContext(DatabaseContext);
    if (!context.database) {
        throw new Error('useDatabase must be used within a DatabaseProvider');
    }
    return context.database;
}
