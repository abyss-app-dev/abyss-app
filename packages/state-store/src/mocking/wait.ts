export function wait(ms = 50) {
    return new Promise(resolve => setTimeout(resolve, ms));
}
