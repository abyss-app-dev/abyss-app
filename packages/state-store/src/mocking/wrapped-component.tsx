import type { StateRecorder } from './state-recorder';

export function TestRecorder<T>({ functionToTest, recorder }: { functionToTest: () => T; recorder: StateRecorder }) {
    recorder.record(functionToTest());
    return <>content</>;
}
