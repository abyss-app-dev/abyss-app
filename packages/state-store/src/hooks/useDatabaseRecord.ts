import type { SqliteTableRecordType, SqliteTables } from '@abyss/records';
import { useEffect } from 'react';
import { useDatabase } from '../context/DatabaseContext';
import { useQuery } from './useQuery';

export function useDatabaseRecord<T extends keyof SqliteTables>(table: T, recordId: string | undefined) {
    const database = useDatabase();

    const query = useQuery(async () => {
        if (!recordId) return null;
        return database.tables[table].get(recordId) as Promise<SqliteTableRecordType[T] | null>;
    }, [table, recordId]);

    useEffect(() => {
        if (!recordId) return;

        let unsubscribe: () => void = () => {};

        database.tables[table]
            .subscribeRecord(recordId, data => {
                query.setData(data as SqliteTableRecordType[T] | null);
            })
            .then(unsub => {
                unsubscribe = unsub;
            });

        return () => unsubscribe();
    }, [database, table, recordId, query.setData]);

    return query;
}
