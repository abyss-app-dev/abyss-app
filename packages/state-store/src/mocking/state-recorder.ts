import { render } from '@testing-library/react';
import type { ReactNode } from 'react';
import { expect } from 'vitest';
import { wait } from './wait';

export class StateRecorder {
    private states: unknown[] = [];

    record(state: unknown): void {
        this.states.push(state);
    }

    async delayedResolve(): Promise<unknown[]> {
        await wait();
        return this.states;
    }

    getStates(): unknown[] {
        return this.states;
    }

    getLatestState(): unknown {
        return this.states[this.states.length - 1];
    }

    clear(): void {
        this.states.length = 0;
    }

    async renderToSnapshot(component: ReactNode): Promise<void> {
        render(component);
        expect(await this.delayedResolve()).toMatchSnapshot();
    }
}
