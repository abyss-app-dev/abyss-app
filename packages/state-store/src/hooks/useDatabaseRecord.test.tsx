import { buildTestDB, type SQliteClient, SqliteTable } from '@abyss/records';
import { useEffect, useState } from 'react';
import { beforeEach, describe, test } from 'vitest';
import { DatabaseProvider } from '../context/DatabaseContext';
import { StateRecorder } from '../mocking/state-recorder';
import { TestRecorder } from '../mocking/wrapped-component';
import { useDatabaseRecord } from './useDatabaseRecord';

describe('useDatabaseRecord hook', () => {
    let db: SQliteClient;
    let recorder: StateRecorder;

    beforeEach(async () => {
        db = await buildTestDB();
        recorder = new StateRecorder();
    });

    test('loads record with valid ID', async () => {
        const defaultRecord = await db.tables[SqliteTable.settings].default();

        await recorder.renderToSnapshot(
            <DatabaseProvider database={db}>
                <TestRecorder
                    recorder={recorder}
                    functionToTest={() => {
                        const result = useDatabaseRecord(SqliteTable.settings, defaultRecord.id);
                        return result;
                    }}
                />
            </DatabaseProvider>
        );
    });

    test('handles null record ID', async () => {
        await recorder.renderToSnapshot(
            <DatabaseProvider database={db}>
                <TestRecorder
                    recorder={recorder}
                    functionToTest={() => {
                        const result = useDatabaseRecord(SqliteTable.settings, undefined);
                        return result;
                    }}
                />
            </DatabaseProvider>
        );
    });

    test('handles non-existent record ID', async () => {
        await recorder.renderToSnapshot(
            <DatabaseProvider database={db}>
                <TestRecorder
                    recorder={recorder}
                    functionToTest={() => {
                        const result = useDatabaseRecord(SqliteTable.settings, 'settings::nonexistent');
                        return result;
                    }}
                />
            </DatabaseProvider>
        );
    });

    test('record ID changes dynamically', async () => {
        const defaultRecord = await db.tables[SqliteTable.settings].default();

        await recorder.renderToSnapshot(
            <DatabaseProvider database={db}>
                <TestRecorder
                    recorder={recorder}
                    functionToTest={() => {
                        const [recordId, setRecordId] = useState<string | undefined>(undefined);

                        const result = useDatabaseRecord(SqliteTable.settings, recordId);

                        useEffect(() => {
                            if (!recordId) {
                                setTimeout(() => setRecordId(defaultRecord.id));
                            }
                        }, [recordId]);

                        return { ...result, recordId };
                    }}
                />
            </DatabaseProvider>
        );
    });

    test('manual refetch functionality', async () => {
        const defaultRecord = await db.tables[SqliteTable.settings].default();

        await recorder.renderToSnapshot(
            <DatabaseProvider database={db}>
                <TestRecorder
                    recorder={recorder}
                    functionToTest={() => {
                        const [hasRefetched, setHasRefetched] = useState(false);

                        const result = useDatabaseRecord(SqliteTable.settings, defaultRecord.id);

                        useEffect(() => {
                            if (result.data && !result.loading && !hasRefetched) {
                                setHasRefetched(true);
                                setTimeout(() => result.refetch());
                            }
                        }, [result.data, result.loading, result.refetch, hasRefetched]);

                        return { ...result, hasRefetched };
                    }}
                />
            </DatabaseProvider>
        );
    });

    test('subscription receives updates', async () => {
        const defaultRecord = await db.tables[SqliteTable.settings].default();

        await recorder.renderToSnapshot(
            <DatabaseProvider database={db}>
                <TestRecorder
                    recorder={recorder}
                    functionToTest={() => {
                        const [hasSetData, setHasSetData] = useState(false);

                        const result = useDatabaseRecord(SqliteTable.settings, defaultRecord.id);

                        useEffect(() => {
                            if (result.data && !result.loading && !hasSetData && result.setData) {
                                setHasSetData(true);
                                setTimeout(() =>
                                    result.setData({
                                        ...defaultRecord,
                                        theme: 'updated-theme',
                                    })
                                );
                            }
                        }, [result.data, result.loading, result.setData, hasSetData]);

                        return { ...result, hasSetData };
                    }}
                />
            </DatabaseProvider>
        );
    });

    test('subscription updates when database record is modified', async () => {
        const defaultRecord = await db.tables[SqliteTable.settings].default();

        await recorder.renderToSnapshot(
            <DatabaseProvider database={db}>
                <TestRecorder
                    recorder={recorder}
                    functionToTest={() => {
                        const [hasUpdatedDb, setHasUpdatedDb] = useState(false);
                        const [updateCount, setUpdateCount] = useState(0);

                        const result = useDatabaseRecord(SqliteTable.settings, defaultRecord.id);

                        // Track when data changes to count subscription updates
                        useEffect(() => {
                            if (result.data) {
                                setUpdateCount(prev => prev + 1);
                            }
                        }, [result.data]);

                        // Modify the database record after initial load
                        useEffect(() => {
                            if (result.data && !result.loading && !hasUpdatedDb) {
                                setHasUpdatedDb(true);
                                setTimeout(async () => {
                                    // Create a new settings record that should trigger subscription
                                    const newRecord = await db.tables[SqliteTable.settings].default();
                                    // This will trigger database updates
                                }, 50);
                            }
                        }, [result.data, result.loading, hasUpdatedDb]);

                        return {
                            ...result,
                            hasUpdatedDb,
                            updateCount,
                            currentTheme: result.data?.theme,
                        };
                    }}
                />
            </DatabaseProvider>
        );
    });

    test('subscription tracks multiple record state changes', async () => {
        const defaultRecord = await db.tables[SqliteTable.settings].default();

        await recorder.renderToSnapshot(
            <DatabaseProvider database={db}>
                <TestRecorder
                    recorder={recorder}
                    functionToTest={() => {
                        const [modifications, setModifications] = useState<string[]>([]);
                        const [step, setStep] = useState(0);

                        const result = useDatabaseRecord(SqliteTable.settings, defaultRecord.id);

                        // Track each data change
                        useEffect(() => {
                            if (result.data) {
                                setModifications(prev => [...prev, `loaded-${result.data?.theme}`]);
                            } else if (result.data === null && !result.loading) {
                                setModifications(prev => [...prev, 'null-state']);
                            }
                        }, [result.data, result.loading]);

                        // Perform a series of manual state changes to test subscription
                        useEffect(() => {
                            if (result.data && !result.loading && step === 0) {
                                setStep(1);
                                setTimeout(() => {
                                    if (result.setData) {
                                        result.setData({
                                            ...defaultRecord,
                                            theme: 'manual-update-1',
                                        });
                                    }
                                }, 50);
                            } else if (step === 1) {
                                setStep(2);
                                setTimeout(() => {
                                    if (result.setData) {
                                        result.setData({
                                            ...defaultRecord,
                                            theme: 'manual-update-2',
                                        });
                                    }
                                }, 50);
                            }
                        }, [result.data, result.loading, result.setData, step]);

                        return {
                            ...result,
                            modifications,
                            step,
                            modificationsCount: modifications.length,
                        };
                    }}
                />
            </DatabaseProvider>
        );
    });
});
