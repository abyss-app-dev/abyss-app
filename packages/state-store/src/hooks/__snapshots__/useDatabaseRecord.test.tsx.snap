// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`useDatabaseRecord hook > handles non-existent record ID 1`] = `
[
  {
    "data": null,
    "error": null,
    "loading": true,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": null,
    "error": [Error: Record with id settings::nonexistent not found in table settings],
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
  },
]
`;

exports[`useDatabaseRecord hook > handles null record ID 1`] = `
[
  {
    "data": null,
    "error": null,
    "loading": true,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": null,
    "error": null,
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
  },
]
`;

exports[`useDatabaseRecord hook > loads record with valid ID 1`] = `
[
  {
    "data": null,
    "error": null,
    "loading": true,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": {
      "createdAt": 1748749550394,
      "id": "settings::default",
      "lastPage": "/",
      "theme": "",
      "updatedAt": 1748749550394,
    },
    "error": null,
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": {
      "createdAt": 1748749550394,
      "id": "settings::default",
      "lastPage": "/",
      "theme": "",
      "updatedAt": 1748749550394,
    },
    "error": null,
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
  },
]
`;

exports[`useDatabaseRecord hook > manual refetch functionality 1`] = `
[
  {
    "data": null,
    "error": null,
    "hasRefetched": false,
    "loading": true,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": {
      "createdAt": 1748749551463,
      "id": "settings::default",
      "lastPage": "/",
      "theme": "",
      "updatedAt": 1748749551463,
    },
    "error": null,
    "hasRefetched": false,
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": {
      "createdAt": 1748749551463,
      "id": "settings::default",
      "lastPage": "/",
      "theme": "",
      "updatedAt": 1748749551463,
    },
    "error": null,
    "hasRefetched": true,
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": {
      "createdAt": 1748749551463,
      "id": "settings::default",
      "lastPage": "/",
      "theme": "",
      "updatedAt": 1748749551463,
    },
    "error": null,
    "hasRefetched": true,
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": {
      "createdAt": 1748749551463,
      "id": "settings::default",
      "lastPage": "/",
      "theme": "",
      "updatedAt": 1748749551463,
    },
    "error": null,
    "hasRefetched": true,
    "loading": true,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": {
      "createdAt": 1748749551463,
      "id": "settings::default",
      "lastPage": "/",
      "theme": "",
      "updatedAt": 1748749551463,
    },
    "error": null,
    "hasRefetched": true,
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
  },
]
`;

exports[`useDatabaseRecord hook > record ID changes dynamically 1`] = `
[
  {
    "data": null,
    "error": null,
    "loading": true,
    "recordId": undefined,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": null,
    "error": null,
    "loading": false,
    "recordId": undefined,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": null,
    "error": null,
    "loading": false,
    "recordId": "settings::default",
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": null,
    "error": null,
    "loading": true,
    "recordId": "settings::default",
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": {
      "createdAt": 1748749551196,
      "id": "settings::default",
      "lastPage": "/",
      "theme": "",
      "updatedAt": 1748749551196,
    },
    "error": null,
    "loading": false,
    "recordId": "settings::default",
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": {
      "createdAt": 1748749551196,
      "id": "settings::default",
      "lastPage": "/",
      "theme": "",
      "updatedAt": 1748749551196,
    },
    "error": null,
    "loading": false,
    "recordId": "settings::default",
    "refetch": [Function],
    "setData": [Function],
  },
]
`;

exports[`useDatabaseRecord hook > subscription receives updates 1`] = `
[
  {
    "data": null,
    "error": null,
    "hasSetData": false,
    "loading": true,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": {
      "createdAt": 1748749551734,
      "id": "settings::default",
      "lastPage": "/",
      "theme": "",
      "updatedAt": 1748749551734,
    },
    "error": null,
    "hasSetData": false,
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": {
      "createdAt": 1748749551734,
      "id": "settings::default",
      "lastPage": "/",
      "theme": "",
      "updatedAt": 1748749551734,
    },
    "error": null,
    "hasSetData": true,
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": {
      "createdAt": 1748749551734,
      "id": "settings::default",
      "lastPage": "/",
      "theme": "",
      "updatedAt": 1748749551734,
    },
    "error": null,
    "hasSetData": true,
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": {
      "createdAt": 1748749551734,
      "id": "settings::default",
      "lastPage": "/",
      "theme": "updated-theme",
      "updatedAt": 1748749551734,
    },
    "error": null,
    "hasSetData": true,
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
  },
]
`;

exports[`useDatabaseRecord hook > subscription tracks multiple record state changes 1`] = `
[
  {
    "data": null,
    "error": null,
    "loading": true,
    "modifications": [],
    "modificationsCount": 0,
    "refetch": [Function],
    "setData": [Function],
    "step": 0,
  },
  {
    "data": {
      "createdAt": 1748749552256,
      "id": "settings::default",
      "lastPage": "/",
      "theme": "",
      "updatedAt": 1748749552256,
    },
    "error": null,
    "loading": false,
    "modifications": [],
    "modificationsCount": 0,
    "refetch": [Function],
    "setData": [Function],
    "step": 0,
  },
  {
    "data": {
      "createdAt": 1748749552256,
      "id": "settings::default",
      "lastPage": "/",
      "theme": "",
      "updatedAt": 1748749552256,
    },
    "error": null,
    "loading": false,
    "modifications": [
      "loaded-",
    ],
    "modificationsCount": 1,
    "refetch": [Function],
    "setData": [Function],
    "step": 1,
  },
  {
    "data": {
      "createdAt": 1748749552256,
      "id": "settings::default",
      "lastPage": "/",
      "theme": "",
      "updatedAt": 1748749552256,
    },
    "error": null,
    "loading": false,
    "modifications": [
      "loaded-",
    ],
    "modificationsCount": 1,
    "refetch": [Function],
    "setData": [Function],
    "step": 2,
  },
  {
    "data": {
      "createdAt": 1748749552256,
      "id": "settings::default",
      "lastPage": "/",
      "theme": "",
      "updatedAt": 1748749552256,
    },
    "error": null,
    "loading": false,
    "modifications": [
      "loaded-",
    ],
    "modificationsCount": 1,
    "refetch": [Function],
    "setData": [Function],
    "step": 2,
  },
  {
    "data": {
      "createdAt": 1748749552256,
      "id": "settings::default",
      "lastPage": "/",
      "theme": "",
      "updatedAt": 1748749552256,
    },
    "error": null,
    "loading": false,
    "modifications": [
      "loaded-",
      "loaded-",
    ],
    "modificationsCount": 2,
    "refetch": [Function],
    "setData": [Function],
    "step": 2,
  },
]
`;

exports[`useDatabaseRecord hook > subscription updates when database record is modified 1`] = `
[
  {
    "currentTheme": undefined,
    "data": null,
    "error": null,
    "hasUpdatedDb": false,
    "loading": true,
    "refetch": [Function],
    "setData": [Function],
    "updateCount": 0,
  },
  {
    "currentTheme": "",
    "data": {
      "createdAt": 1748749551995,
      "id": "settings::default",
      "lastPage": "/",
      "theme": "",
      "updatedAt": 1748749551995,
    },
    "error": null,
    "hasUpdatedDb": false,
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
    "updateCount": 0,
  },
  {
    "currentTheme": "",
    "data": {
      "createdAt": 1748749551995,
      "id": "settings::default",
      "lastPage": "/",
      "theme": "",
      "updatedAt": 1748749551995,
    },
    "error": null,
    "hasUpdatedDb": true,
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
    "updateCount": 1,
  },
  {
    "currentTheme": "",
    "data": {
      "createdAt": 1748749551995,
      "id": "settings::default",
      "lastPage": "/",
      "theme": "",
      "updatedAt": 1748749551995,
    },
    "error": null,
    "hasUpdatedDb": true,
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
    "updateCount": 1,
  },
  {
    "currentTheme": "",
    "data": {
      "createdAt": 1748749551995,
      "id": "settings::default",
      "lastPage": "/",
      "theme": "",
      "updatedAt": 1748749551995,
    },
    "error": null,
    "hasUpdatedDb": true,
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
    "updateCount": 2,
  },
]
`;
